### ISSUE TITLE
Occasionally hang on unlocking a document

### ISSUE THREAD

**Issue by 1024jp** [coteditor/CotEditor#558]
It works, but something's going wrong.

```
2016-05-18 19:31:01.907 CotEditor[66131:5058457] -continueFileAccessUsingBlock: was invoked without an outer file access! File access performed inside this block is not safe! Here's the backtrace: (
	0   AppKit                              0x00007fff89cafb1a -[NSDocument _continueCurrentFileAccessDuringBlock:] + 331
	1   AppKit                              0x00007fff89c62c17 -[NSDocument performSynchronousFileAccessUsingBlock:] + 117
	2   AppKit                              0x00007fff89ce65f1 -[NSDocument _checkAutosavingThenContinue:] + 296
	3   AppKit                              0x00007fff89e16709 __50-[NSDocument unlockDocumentWithCompletionHandler:]_block_invoke_2 + 554
	4   AppKit                              0x00007fff89e16aa2 __42-[NSDocument unlockWithCompletionHandler:]_block_invoke + 108
	5   AppKit                              0x00007fff89b2fd27 -[NSDocument continueFileAccessUsingBlock:] + 222
	6   AppKit                              0x00007fff89df1a4b __33-[NSDocument _performFileAccess:]_block_invoke769 + 125
	7   AppKit                              0x00007fff89e3a883 __62-[NSDocumentController(NSInternal) _onMainThreadInvokeWorker:]_block_invoke2079 + 175
	8   CoreFoundation                      0x00007fff8fa684ac __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
	9   CoreFoundation                      0x00007fff8fa47d05 __CFRunLoopDoBlocks + 341
	10  CoreFoundation                      0x00007fff8fa47a6b __CFRunLoopRun + 2347
	11  CoreFoundation                      0x00007fff8fa46ed8 CFRunLoopRunSpecific + 296
	12  HIToolbox                           0x00007fff994d8935 RunCurrentEventLoopInMode + 235
	13  HIToolbox                           0x00007fff994d876f ReceiveNextEventCommon + 432
	14  HIToolbox                           0x00007fff994d85af _BlockUntilNextEventMatchingListInModeWithFilter + 71
	15  AppKit                              0x00007fff898d8df6 _DPSNextEvent + 1067
	16  AppKit                              0x00007fff898d8226 -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454
	17  AppKit                              0x00007fff898ccd80 -[NSApplication run] + 682
	18  AppKit                              0x00007fff89896368 NSApplicationMain + 1176
	19  CotEditor                           0x0000000100001b42 main + 34
	20  libdyld.dylib                       0x00007fff8a6ac5ad start + 1
)
```

**Comment by 1024jp** [comment]
The application can even hang on unlocking. However, the similar thing occurs also on TextEdit.app. So this can be not CotEditor-related matter.

**Comment by 1024jp** [comment]
Hang log.

```
Date/Time:       2016-05-18 22:48:06 +0900
OS Version:      Mac OS X 10.11.5 (Build 15F34)
Architecture:    x86_64h
Report Version:  22

Command:         CotEditor
Path:            /Applications/CotEditor.app/Contents/MacOS/CotEditor
Version:         2.5.4 (127)
App Item ID:     1024640650
App External ID: 817370395
Parent:          launchd [1]
PID:             70784

Event:           hang
Duration:        2.00s (process was unresponsive for 11 seconds before sampling)
Steps:           20 (100ms sampling interval)

Hardware model:  MacBookPro12,1
Active cpus:     4

Fan speed:       0 rpm

--------------------------------------------------
Timeline format: stacks are sorted chronologically
Use -i and -heavy to re-report with count sorting
--------------------------------------------------

Heaviest stack for the main thread of the target process:
  20  start + 1 (libdyld.dylib + 13741) [0x7fff8a6ac5ad]
  20  NSApplicationMain + 1176 (AppKit + 25448) [0x7fff89896368]
  20  -[NSApplication run] + 682 (AppKit + 249216) [0x7fff898ccd80]
  20  -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454 (AppKit + 295462) [0x7fff898d8226]
  20  _DPSNextEvent + 1067 (AppKit + 298486) [0x7fff898d8df6]
  20  _BlockUntilNextEventMatchingListInModeWithFilter + 71 (HIToolbox + 198063) [0x7fff994d85af]
  20  ReceiveNextEventCommon + 432 (HIToolbox + 198511) [0x7fff994d876f]
  20  RunCurrentEventLoopInMode + 235 (HIToolbox + 198965) [0x7fff994d8935]
  20  CFRunLoopRunSpecific + 296 (CoreFoundation + 560856) [0x7fff8fa46ed8]
  20  __CFRunLoopRun + 2347 (CoreFoundation + 563819) [0x7fff8fa47a6b]
  20  __CFRunLoopDoBlocks + 341 (CoreFoundation + 564485) [0x7fff8fa47d05]
  20  __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12 (CoreFoundation + 697516) [0x7fff8fa684ac]
  20  __62-[NSDocumentController(NSInternal) _onMainThreadInvokeWorker:]_block_invoke2079 + 175 (AppKit + 5941379) [0x7fff89e3a883]
  20  __33-[NSDocument _performFileAccess:]_block_invoke769 + 125 (AppKit + 5642827) [0x7fff89df1a4b]
  20  -[NSDocument continueFileAccessUsingBlock:] + 222 (AppKit + 2751783) [0x7fff89b2fd27]
  20  __42-[NSDocument unlockWithCompletionHandler:]_block_invoke + 108 (AppKit + 5794466) [0x7fff89e16aa2]
  20  __50-[NSDocument unlockDocumentWithCompletionHandler:]_block_invoke_2 + 554 (AppKit + 5793545) [0x7fff89e16709]
  20  -[NSDocument _checkAutosavingThenContinue:] + 296 (AppKit + 4548081) [0x7fff89ce65f1]
  20  -[NSDocument performSynchronousFileAccessUsingBlock:] + 117 (AppKit + 4008983) [0x7fff89c62c17]
  20  -[NSDocument _continueCurrentFileAccessDuringBlock:] + 205 (AppKit + 4323996) [0x7fff89cafa9c]
  20  semaphore_wait_trap + 10 (libsystem_kernel.dylib + 69550) [0x7fff8cd1ffae]
 *20  semaphore_wait_continue + 0 (kernel + 1029504) [0xffffff80002fb580]


Process:         CotEditor [70784]
Path:            /Applications/CotEditor.app/Contents/MacOS/CotEditor
Architecture:    x86_64
Parent:          launchd [1]
UID:             501
Task size:       19724 pages
CPU Time:        0.004s
Note:            Unresponsive for 11 seconds before sampling
Note:            2 idle work queue threads omitted

  Thread 0x51b666         DispatchQueue 1         20 samples (1-20)       priority 47 (base 47)
  <process frontmost, thread QoS user interactive (requested user interactive), process unclamped, process received importance donation from WindowServer [201], process received importance donation from DocumentPopoverViewService [71423], process received importance donation from ViewBridgeAuxiliary [2618], IO policy important>
  20  start + 1 (libdyld.dylib + 13741) [0x7fff8a6ac5ad] 1-20
    20  NSApplicationMain + 1176 (AppKit + 25448) [0x7fff89896368] 1-20
      20  -[NSApplication run] + 682 (AppKit + 249216) [0x7fff898ccd80] 1-20
        20  -[NSApplication _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 454 (AppKit + 295462) [0x7fff898d8226] 1-20
          20  _DPSNextEvent + 1067 (AppKit + 298486) [0x7fff898d8df6] 1-20
            20  _BlockUntilNextEventMatchingListInModeWithFilter + 71 (HIToolbox + 198063) [0x7fff994d85af] 1-20
              20  ReceiveNextEventCommon + 432 (HIToolbox + 198511) [0x7fff994d876f] 1-20
                20  RunCurrentEventLoopInMode + 235 (HIToolbox + 198965) [0x7fff994d8935] 1-20
                  20  CFRunLoopRunSpecific + 296 (CoreFoundation + 560856) [0x7fff8fa46ed8] 1-20
                    20  __CFRunLoopRun + 2347 (CoreFoundation + 563819) [0x7fff8fa47a6b] 1-20
                      20  __CFRunLoopDoBlocks + 341 (CoreFoundation + 564485) [0x7fff8fa47d05] 1-20
                        20  __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12 (CoreFoundation + 697516) [0x7fff8fa684ac] 1-20
                          20  __62-[NSDocumentController(NSInternal) _onMainThreadInvokeWorker:]_block_invoke2079 + 175 (AppKit + 5941379) [0x7fff89e3a883] 1-20
                            20  __33-[NSDocument _performFileAccess:]_block_invoke769 + 125 (AppKit + 5642827) [0x7fff89df1a4b] 1-20
                              20  -[NSDocument continueFileAccessUsingBlock:] + 222 (AppKit + 2751783) [0x7fff89b2fd27] 1-20
                                20  __42-[NSDocument unlockWithCompletionHandler:]_block_invoke + 108 (AppKit + 5794466) [0x7fff89e16aa2] 1-20
                                  20  __50-[NSDocument unlockDocumentWithCompletionHandler:]_block_invoke_2 + 554 (AppKit + 5793545) [0x7fff89e16709] 1-20
                                    20  -[NSDocument _checkAutosavingThenContinue:] + 296 (AppKit + 4548081) [0x7fff89ce65f1] 1-20
                                      20  -[NSDocument performSynchronousFileAccessUsingBlock:] + 117 (AppKit + 4008983) [0x7fff89c62c17] 1-20
                                        20  -[NSDocument _continueCurrentFileAccessDuringBlock:] + 205 (AppKit + 4323996) [0x7fff89cafa9c] 1-20
                                          20  semaphore_wait_trap + 10 (libsystem_kernel.dylib + 69550) [0x7fff8cd1ffae] 1-20
                                           *20  semaphore_wait_continue + 0 (kernel + 1029504) [0xffffff80002fb580] 1-20

  Thread 0x51b68e         DispatchQueue 2         20 samples (1-20)       priority 47 (base 47)
  <process frontmost, thread QoS user interactive (requested user interactive), process unclamped, process received importance donation from WindowServer [201], process received importance donation from DocumentPopoverViewService [71423], process received importance donation from ViewBridgeAuxiliary [2618], IO policy important>
  20  _dispatch_mgr_thread + 52 (libdispatch.dylib + 32205) [0x7fff92f42dcd] 1-20
    20  kevent_qos + 10 (libsystem_kernel.dylib + 98042) [0x7fff8cd26efa] 1-20
     *20  ??? (kernel + 5828224) [0xffffff800078ee80] 1-20

  Thread 0x51b746         20 samples (1-20)       priority 47 (base 47)   cpu time 0.004s
  <process frontmost, thread QoS user interactive (requested user interactive), process unclamped, process received importance donation from WindowServer [201], process received importance donation from DocumentPopoverViewService [71423], process received importance donation from ViewBridgeAuxiliary [2618], IO policy important>
  20  thread_start + 13 (libsystem_pthread.dylib + 4945) [0x7fff92f6b351] 1-20
    20  _pthread_start + 168 (libsystem_pthread.dylib + 14618) [0x7fff92f6d91a] 1-20
      20  _pthread_body + 131 (libsystem_pthread.dylib + 14749) [0x7fff92f6d99d] 1-20
        20  _NSEventThread + 149 (AppKit + 1699221) [0x7fff89a2ed95] 1-20
          20  CFRunLoopRunSpecific + 296 (CoreFoundation + 560856) [0x7fff8fa46ed8] 1-20
            20  __CFRunLoopRun + 1356 (CoreFoundation + 562828) [0x7fff8fa4768c] 1-20
              20  __CFRunLoopServiceMachPort + 212 (CoreFoundation + 565700) [0x7fff8fa481c4] 1-20
                20  mach_msg_trap + 10 (libsystem_kernel.dylib + 69490) [0x7fff8cd1ff72] 1-20
                 *20  ipc_mqueue_receive_continue + 0 (kernel + 812624) [0xffffff80002c6650] 1-20

  Binary Images:
         0x1042b0000 -        0x104398fff  com.coteditor.CotEditor 2.5.4 (127)   <10B0D25B-EB63-3293-A391-921E22734616>  /Applications/CotEditor.app/Contents/MacOS/CotEditor
      0x7fff89890000 -     0x7fff8a4b9fff  com.apple.AppKit 6.9 (1404.47)        <F3411F6E-DD87-34D0-8C68-C69B2205E41D>  /System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
      0x7fff8a6a9000 -     0x7fff8a6acfff  libdyld.dylib (360.22)                <CC088C2A-D407-33E7-A6B6-B06E0D4AD999>  /usr/lib/system/libdyld.dylib
      0x7fff8cd0f000 -     0x7fff8cd2dfff  libsystem_kernel.dylib (3248.50.21)   <78E54D59-D2B0-3F54-9A4A-0A68D671F253>  /usr/lib/system/libsystem_kernel.dylib
      0x7fff8f9be000 -     0x7fff8fe34fff  com.apple.CoreFoundation 6.9 (1258.1) <943A1383-DA6A-3DC0-ABCD-D9AEB3D0D34D>  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
      0x7fff92f3b000 -     0x7fff92f68fff  libdispatch.dylib (501.40.12)         <C7499857-61A5-3D7D-A5EA-65DCC8C3DF92>  /usr/lib/system/libdispatch.dylib
      0x7fff92f6a000 -     0x7fff92f73fff  libsystem_pthread.dylib (138.10.4)    <3DD1EF4C-1D1B-3ABF-8CC6-B3B1CEEE9559>  /usr/lib/system/libsystem_pthread.dylib
      0x7fff994a8000 -     0x7fff9979dfff  com.apple.HIToolbox 2.1.1 (807.2)     <36413C45-36AF-34EF-9C0E-F18B31D1E565>  /System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/HIToolbox
 *0xffffff8000200000 - 0xffffff80009fffff  kernel (3248.50.21)                   <7E7B0822-D2DE-3B39-A7A5-77B40A668BC6>  /System/Library/Kernels/kernel
```

**Comment by handcoding** [comment]
I’m not sure whether this might be the same issue, but after I upgraded to CotEditor 2.5.5 and I started it up for the first time, it hung while restoring my windows.

If it might help, I’ve also attached a “Sample Process” log which I captured during CotEditor’s hang. 

[CotEditor 2-5-5 - hang when restoring windows.txt](https://github.com/coteditor/CotEditor/files/277745/CotEditor.2-5-5.-.hang.when.restoring.windows.txt)

**Comment by 1024jp** [comment]
Hmmm, thank your for the hang log.

Is this issue occurred only once after the update or happens every time app launched?
Seeing your hang log, I suppose it isn't related to the issue I mentioned.
I'm not sure. But if it occurs just only once and can not be reproduced any more, I think yours isn't something to be fixed.

**Comment by handcoding** [comment]
> Is this issue occurred only once after the update or happens every time app launched?

After I force-quit CotEditor during its hang, the next time I started CotEditor, it gave me an option to “Don’t Reopen” or “Reopen” windows: 
![coteditor 2-5-5 - hang when restoring windows](https://cloud.githubusercontent.com/assets/118319/15476908/8c8cdffe-20d7-11e6-9134-b7de8f936d6c.png)

So I chose “Reopen”, but then CotEditor hung again. So I force-quit CotEditor again, and then the next time I started it, I chose “Don’t Reopen.” At that point, I manually reopened my previously open files (or as many as I could remember having open), and CotEditor seemed to be okay.

For good measure, I then quit CotEditor and then reopened it—and CotEditor was able to reopen those files. To be sure, I didn’t have as many files open compared to when CotEditor hung (because I couldn’t remember every last file that I had open previously), so I wonder whether CotEditor’s earlier hangs might somehow be related to the number of files that it’s trying to reopen at one time?

**Comment by handcoding** [comment]
> I'm not sure. But if it occurs just only once and can not be reproduced any more, I think yours isn't something to be fixed.

For what it’s worth, CotEditor 2.5.6 hangs on startup every time I reboot my machine. (I have CotEditor set to open at login.) When I then force-quit CotEditor and restart it, I get the dialog from my previous comment—the one that says, “The last time you opened CotEditor, it was force quite while reopening windows. Do you want to try to reopen its windows again?”

From there, if I choose “Don’t Reopen”, CotEditor will open normally (although without reopening any of my files from last time). 

If it may help, I’ve also attached a “Sample Process” log from Activity Manager that I captured during CotEditor’s most recent startup hang.

[CotEditor - Activity Monitor sample during startup - 2016-07-19.txt](https://github.com/coteditor/CotEditor/files/372543/CotEditor.-.Activity.Monitor.sample.during.startup.-.2016-07-19.txt)

**Comment by 1024jp** [comment]
Thank you for continuous reports.

By reading your report and log right now, I remember that I faced to a similar phenomenon while I developing CotEditor 3. I thought at first it is something related to the Swift migration, however actually it can be related to your issue. If my guess is right, I probably can solve the issue also on CotEditor 2.5.x. I'm still not sure but it seems worth trying.

Recently I'm planning to release CotEditor 2.5.7 soon even there are only trivial fixes. So I'll contain also the fix for this to 2.5.7.

**Comment by handcoding** [comment]
> Recently I'm planning to release CotEditor 2.5.7 soon even there are only trivial fixes. So I'll contain also the fix for this to 2.5.7.

Oh, that’s fantastic news! (Yay!)

P.S. As a side thing, “recently” is typically only used for events in the recent _past_. For events in the near future, it would be more common to use words like “soon” or “shortly.” (Or if by chance these sorts of comments about English are unwanted, feel free to tell me.)

**Comment by 1024jp** [comment]
CotEditor 3.0 changes a lot and this original issue seems to be already resolved.
So, I close this issue. If you still get crash on 3.0, please create another issue.

> P.S. As a side thing, “recently” is typically only used for events in the recent past. For events in the near future, it would be more common to use words like “soon” or “shortly.” (Or if by chance these sorts of comments about English are unwanted, feel free to tell me.)

Oops, sorry for my late reply. I missed your mention.
Thank you for the correction and I welcome to such a pointing-out. I know I take less care with grammar when I write something on issue tracker ;-P

**Comment by kisileno** [comment]
Hi @1024jp, could you please advise how do you fixed the issue from code/project settings perspective?

**Comment by 1024jp** [comment]
Hi @kisileno, are you facing a simliar problem?

Honestly, I'm still not 100% sure how it works and don't remember well how I fixed this at that moment. I detoured this just through trial and error. I've tried a lot for this file access lock issue on NSDocument.

My guess is that something is wrong with the async saving on NSDocument. Previously, Apple addressed this issue on OS X 10.11 ([NSDocument Deadlock Avoidance](https://developer.apple.com/library/prerelease/content/releasenotes/AppKit/RN-AppKitOlderNotes/index.html#X10_11Notes)), however, it seems some issues remain still. Probably, the best is to return `false` in `canAsynchronouslyWrite(to:ofType:for:)`, but, you know, it cause an impact on application performance. So, better not.

I've tried a lot for this file access lock issue on NSDocument. My current strategy agains this issue is:

- Minimize the access to the document file, including an access through a NSFileCoordinator.
- Give up playing with `performAsynchronousFileAccess()` series. They don't work as their description says.
- The source code of TextEdit.app, the sample sorce code by Apple for the document-based Cocoa app, is too old and now unreliable in large parts.

Or, if you talking about the crash/hang on document reopen, I reccomend  you to check over the code around `restoreState(with:)`.

Sorry for my unclear response. There is no enough information about document-based Cocoa app to clarify this issue.

**Comment by kisileno** [comment]
@1024jp Unfortunately, I have **reproduced the issue** in CotEdtior from AppStore, TextEdit, Pages and my NSDocument-based app on latest El C*r*apitan. Steps to reproduce are pretty straightforward: just spam `lock/unlock checkbox` till program hangs (4-5 clicks).  
Thank you for  comprehensive response, will play with `sync` persistence. Also, I am looking is it possible to turn off this functionality at all.

**Comment by 1024jp** [comment]
@kisileno I see. I probably confound with several simliar issues when I answered your first question. Hmm, this problem is definitely the type that Apple should fix (I know, even so, we developers need to fix it sometimes). 

Fortunately, it seems this issue was already resolved on macOS 10.12.  I actually didn't know even that this problem still remains in my CotEditor since I've updated my main computer to macOS 10.12. I tried again to reproduce the issue in CotEditor and also Pages on macOS 10.12, but I couldn't. (Therefore, in my case, I'll let this be 😜)

I don't know if there is a way to fix it on 10.11. But anyway you should check yours on macOS 10.12.
